akka-typed/src/main/scala/akka/typed/internal/Restarter.scala

```
@@ -0,0 +1,272 @@
+  def apply[T, Thr <: Throwable: ClassTag](initialBehavior: Behavior[T], strategy: SupervisorStrategy): Behavior[T] =
+    Actor.Deferred[T] { ctx ⇒
+      val c = ctx.asInstanceOf[akka.typed.ActorContext[T]]
+      val startedBehavior = initialUndefer(ctx.asInstanceOf[akka.typed.ActorContext[T]], initialBehavior)
+      strategy match {
+        case Restart(-1, _, loggingEnabled) ⇒
+          new Restarter(initialBehavior, startedBehavior, loggingEnabled)
+        case r: Restart ⇒
+          new LimitedRestarter(initialBehavior, startedBehavior, r, retries = 0, deadline = OptionVal.None)
+        case Resume(loggingEnabled) ⇒ new Resumer(startedBehavior, loggingEnabled)
+        case b: Backoff ⇒
+          val backoffRestarter =
+            new BackoffRestarter(
+              initialBehavior.asInstanceOf[Behavior[Any]],
+              startedBehavior.asInstanceOf[Behavior[Any]],
+              b, restartCount = 0, blackhole = false)
+          backoffRestarter.asInstanceOf[Behavior[T]]
+      }
+    }
+
+  def initialUndefer[T](ctx: ActorContext[T], initialBehavior: Behavior[T]): Behavior[T] =
+    Behavior.validateAsInitial(Behavior.undefer(initialBehavior, ctx))
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] abstract class Supervisor[T, Thr <: Throwable: ClassTag] extends ExtensibleBehavior[T] {
+
+  protected def loggingEnabled: Boolean
+
+  /**
+   * Current behavior
+   */
+  protected def behavior: Behavior[T]
+
+  /**
+   * Wrap next behavior in a concrete restarter again.
+   */
+  protected def wrap(nextBehavior: Behavior[T], afterException: Boolean): Supervisor[T, Thr]
+
+  protected def handleException(ctx: ActorContext[T], startedBehavior: Behavior[T]): Catcher[Supervisor[T, Thr]]
+
+  protected def restart(ctx: ActorContext[T], initialBehavior: Behavior[T], startedBehavior: Behavior[T]): Supervisor[T, Thr] = {
+    try Behavior.interpretSignal(startedBehavior, ctx, PreRestart) catch { case NonFatal(_) ⇒ }
```

is that swallowing of exception ok here?